src/main/java/uk/gov/homeoffice/digital/sas/timecard/validators/timeentry/TimeEntryValidator.java

```diff
@@ -30,6 +35,87 @@ public boolean isValid(Object value, ConstraintValidatorContext context) {
         timeEntry.getActualEndTime());
 
     session.setHibernateFlushMode(FlushMode.AUTO);
-    return timeEntryClashes.isEmpty();
+    if (!timeEntryClashes.isEmpty()) {
+      ClashingProperty clashingProperty = getClashingProperty(timeEntry, timeEntryClashes);
+      var payload = timeEntryClashes.stream().map(this::transformTimeEntry)
+          .collect(Collectors.toCollection(ArrayList::new));
+
+      HibernateConstraintValidatorContext hibernateContext =
+          context.unwrap(HibernateConstraintValidatorContext.class);
+
+      hibernateContext.disableDefaultConstraintViolation();
+      hibernateContext.withDynamicPayload(payload);
+      hibernateContext
+          .buildConstraintViolationWithTemplate(
+              "Time periods must not overlap with another time period")
+          .addPropertyNode(clashingProperty.toString())
+          .addConstraintViolation();
+      return false;
+    }
+    return true;
+  }
+
+  private JSONObject transformTimeEntry(TimeEntry timeEntry) {
+    var result = new JSONObject();
+    result.put("startTime", timeEntry.getActualStartTime());
+    result.put("endTime", timeEntry.getActualEndTime());
+    result.put("timePeriodTypeId", timeEntry.getTimePeriodTypeId());
+    return result;
+  }
+
+  private ClashingProperty getClashingProperty(
+      TimeEntry timeEntry, List<TimeEntry> timeEntryClashes) {
+    var startTimeClash = false;
+    var endTimeClash = false;
+
+    for (TimeEntry timeEntryClash : timeEntryClashes) {
+      if (timeEntry.getActualStartTime().equals(timeEntryClash.getActualStartTime())) {
```

Comment:
This `if` and the below if could be combined with an `||` and cut down on code.
Unless you have it this way out of preference then feel free to resolve this.

Reply:
I think it's more readable to have them separate (at least for me), but happy to change it if you feel strongly?

Reply:
yeah that's perfectly fine. My first instinct is to just go down the shorter code route aha, but you're right this one is a bit more readable 👍